File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(root == nullptr){
            return nullptr;
        }  
        TreeNode* curr = root;
        while(curr){
            if(p->val > curr->val && q->val > curr->val ){
                curr = curr->right;
            }else if(p->val < curr->val && q->val < curr->val){
                curr = curr->left;
            }else{
                return curr;
            }
        }

        return curr;


        
    }
};

File: Data Structures & Algorithms/balanced-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        
        self.res = True 

        def traverse(curr):
            if not curr:
                return 0

            left_return = traverse(curr.left)
            right_return = traverse(curr.right)

            if abs(left_return - right_return) > 1:
                self.res = False

            return max(left_return, right_return) + 1

        traverse(root)
        return self.res